pkg/cache/sql/db/client.go

```diff
@@ -0,0 +1,312 @@
+/*
+Package db offers client struct and  functions to interact with database connection. It provides encrypting, decrypting,
+and a way to reset the database.
+*/
+package db
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/gob"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/lasso/pkg/cache/sql/attachdriver"
+	"github.com/rancher/lasso/pkg/cache/sql/db/transaction"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"modernc.org/sqlite"
+	sqlite3 "modernc.org/sqlite/lib"
+)
+
+const (
```

Comment:
Are the two databases split so that we can have the `db2` one in memory if encryption is required?

Can that cause any issues with transaction? I couldn't find if SQLite with WAL journal mode supports transaction with multi-databases.

Reply:
@tomleb yes they are split and that is why. The purpose of attach driver is to make sure the second db2 is part of the connection. I haven't read anything that would make me think it would be an issue, especially because we are reusing the connection pool.

Reply:
Yes, that is the reason for db2 running in memory and the purpose of attach driver is to attach every underlying connection started by golang's sql package.  It sounds like the only scenario in which there can be an issue is if if it is WAL and the program crashes. This case does not apply to us since we do not reuse the database between rancher restarts. See https://www.sqlite.org/lang_attach.html:

> Transactions involving multiple attached databases are atomic, assuming that the main database is not "[:memory:](https://www.sqlite.org/inmemorydb.html)" and the [journal_mode](https://www.sqlite.org/pragma.html#pragma_journal_mode) is not [WAL](https://www.sqlite.org/wal.html). If the main database is ":memory:" or if the journal_mode is WAL, then transactions continue to be atomic within each individual database file. But if the host computer crashes in the middle of a [COMMIT](https://www.sqlite.org/lang_transaction.html) where two or more database files are updated, some of those files might get the changes where others might not.